crates/bevy_render/src/render_phase/mod.rs

```diff
@@ -1357,3 +1528,127 @@ impl BinnedRenderPhaseType {
         }
     }
 }
+
+impl RenderBin {
+    /// Creates a [`RenderBin`] containing a single entity.
+    fn from_entity(entity: MainEntity) -> RenderBin {
+        let mut entities = IndexSet::default();
+        entities.insert(entity);
+        RenderBin { entities }
+    }
+
+    /// Inserts an entity into the bin.
+    fn insert(&mut self, entity: MainEntity) {
+        self.entities.insert(entity);
+    }
+
+    /// Removes an entity from the bin.
+    fn remove(&mut self, entity_to_remove: MainEntity) {
+        self.entities.swap_remove(&entity_to_remove);
+    }
+
+    /// Returns true if the bin contains no entities.
+    fn is_empty(&self) -> bool {
+        self.entities.is_empty()
+    }
+
+    /// Returns the [`IndexSet`] containing all the entities in the bin.
+    #[inline]
+    pub fn entities(&self) -> &IndexSet<MainEntity, EntityHash> {
+        &self.entities
+    }
+}
+
+/// An iterator that efficiently finds the indices of all zero bits in a
+/// [`FixedBitSet`] and returns them in reverse order.
+///
+/// [`FixedBitSet`] doesn't natively offer this functionality, so we have to
+/// implement it ourselves.
+#[derive(Debug)]
+struct ReverseFixedBitSetZeroesIterator<'a> {
+    /// The bit set.
+    bitset: &'a FixedBitSet,
+    /// The next bit index we're going to scan when [`Iterator::next`] is
+    /// called.
+    bit_index: isize,
+}
+
+impl<'a> ReverseFixedBitSetZeroesIterator<'a> {
+    fn new(bitset: &'a FixedBitSet) -> ReverseFixedBitSetZeroesIterator<'a> {
+        ReverseFixedBitSetZeroesIterator {
+            bitset,
+            bit_index: (bitset.len() as isize) - 1,
+        }
+    }
+}
+
+impl<'a> Iterator for ReverseFixedBitSetZeroesIterator<'a> {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<usize> {
+        while self.bit_index >= 0 {
+            // Unpack the bit index into block and bit.
+            let block_index = self.bit_index / (Block::BITS as isize);
+            let bit_pos = self.bit_index % (Block::BITS as isize);
+
+            // Grab the block. Mask off all bits before the one we're scanning
```

Comment:
"Before" here caught me up because we're going in reverse. Above? Previous?

Reply:
Changed to "above".